packages/core/src/server/mutation-server-protocol-handler.ts

```diff
@@ -0,0 +1,202 @@
+import {
+  JSONRPCClient,
+  JSONRPCRequest,
+  JSONRPCResponse,
+  JSONRPCServer,
+  JSONRPCServerAndClient,
+  JSONRPCServerMiddlewareNext,
+  TypedJSONRPCServerAndClient,
+  createJSONRPCErrorResponse,
+} from 'json-rpc-2.0';
+import { MutantResult, PartialStrykerOptions } from '@stryker-mutator/api/core';
+import { tokens } from 'typed-inject';
+
+import { MutationTestMethod, InstrumentMethod } from './methods/index.js';
+import { Transporter } from './transport/index.js';
+import * as serverTokens from './server-tokens.js';
+
+import {
+  CancelNotification,
+  ClientMethods,
+  InstrumentParams,
+  MutateParams,
+  MutatePartialResult,
+  ProgressParams,
+  ErrorCodes,
+  ServerMethods,
+  InitializeResult,
+  InitializeParams,
+} from './index.js';
+
+/**
+ * Handles Mutation Server Protocol JSON-RPC messaging using the given transporter.
+ */
+export class MutationServerProtocolHandler {
+  public static readonly inject = tokens(serverTokens.transporter);
+  private readonly serverAndClient: TypedJSONRPCServerAndClient<ServerMethods, ClientMethods>;
+
+  // Map of request id to cancellation callback.
+  private readonly cancellableRequests = new Map<number | string, AbortController>();
+  private readonly strykerOptionsOverwrite: PartialStrykerOptions = {};
+  private initialized = false;
+
+  constructor(private readonly transporter: Transporter) {
+    const jsonRpcServer = new JSONRPCServer();
+    const jsonRpcClient = new JSONRPCClient((message) => transporter.send(JSON.stringify(message)));
+    this.serverAndClient = new JSONRPCServerAndClient(jsonRpcServer, jsonRpcClient);
+
+    this.serverAndClient.applyServerMiddleware(this.serverInitializedMiddleware.bind(this), this.requestCancellationMiddleware.bind(this));
+    this.setupServerMethods();
+    this.setupTransporterCallbacks();
+  }
+
+  private async serverInitializedMiddleware(next: JSONRPCServerMiddlewareNext<void>, request: JSONRPCRequest, serverParams: void) {
+    if (this.initialized && request.method === 'initialize') {
+      // Server is already initialized, only allow initialize requests once.
+      return createJSONRPCErrorResponse(
+        request.id ?? null,
+        ErrorCodes.InvalidRequest,
+        'Server is already initialized, you can only request initialization once.',
+      );
+    } else if (!this.initialized && request.method !== 'initialize') {
+      // Server is not initialized, only allow initialize requests.
+      return createJSONRPCErrorResponse(
+        request.id ?? null,
+        ErrorCodes.ServerNotInitialized,
+        'Server is not initialized, you must request initialization first.',
+      );
+    }
+
+    return next(request, serverParams);
+  }
+
+  private async requestCancellationMiddleware(next: JSONRPCServerMiddlewareNext<void>, request: JSONRPCRequest, serverParams: void) {
+    if (request.id) {
```

Comment:
`id` can be a `number` or a `string`, both of which have falsy values. You probably want an `undefined` check here.